### PR TITLE
Calculate distances using distance maps

### PR DESCRIPTION
**Pull request by huddlej**
Adds support for reading distance map JSONs and calculating distances between
sequences using a given map and the sequences in a given tree. This support
replaces the original mask-based definitions for calculating distances and
provides a more generic way to calculate distances between sequences.

The new distance map interface adds support for three types of comparisons
between:

1. the root and all nodes in the tree (the previous default for all distances)
2. each tip from a current season and its immediate ancestor from a previous season
3. each tip from a current season and all tips in previous seasons

The two new comparisons allow users to specify a "latest date" that defines when
the current season starts. For pairwise comparisons, users can also specify an
"earliest date" that limits how far back in time comparisons should be allowed.

In addition to supporting the standard bit mask-style distance calculations,
this commit also adds support for sequence-specific distance maps (i.e.,
distances for specific amino acid mutations at specific sites). These
sequence-specific maps allow comparisons between nodes using the titer
substitution model and other experimentally defined distances.

Note that this commit maintains the ability for users to request multiple
pairs of attributes and distance maps for a single output file. This
functionality makes integration of the augur distance command into existing
pathogen builds much simpler without making the distance command code itself
more complicated. Users can still run the distance command multiple times, if
they wish.

### CODE REVIEW

File: augur/distance.py [hunk 1]
```diff
@@ -22,6 +24,296 @@ def read_masks(mask_file):
...
+        # Skip nodes that were sampled prior to this date.
+        if latest_date is not None and node.attr["num_date"] < latest_date:
+            continue
+
+        # Distances between this node and other nodes are indexed by the other
+        # node name.
+        distances_by_node[node.name] = {}
+
+        # Find all nodes that were sampled prior to the given latest date.
+        for past_node in tree.find_clades(terminal=True):
```
Comment (tsibley) [inline augur/distance.py]: Should a node's distance to itself be omitted, e.g. when `node.name == past_node.name`?

File: augur/distance.py [hunk 2]
```diff
@@ -51,18 +343,19 @@ def register_arguments(parser):
     parser.add_argument("--tree", help="Newick tree", required=True)
     parser.add_argument("--alignment", nargs="+", help="sequence(s) to be used, supplied as FASTA files", required=True)
     parser.add_argument('--gene-names', nargs="+", type=str, help="names of the sequences in the alignment, same order assumed", required=True)
+    parser.add_argument("--compare-to", nargs="+", choices=["root", "ancestor", "pairwise"], help="type of comparison between samples in the given tree", required=True)
```
Comment (tsibley) [inline augur/distance.py]: Could the help expand upon what each choice means?

File: augur/distance.py [hunk 3]
```diff
@@ -51,18 +343,19 @@ def register_arguments(parser):
...
+    parser.add_argument("--earliest-date", help="earliest date to consider samples for pairwise comparisons (e.g., 2019-01-01); if omitted, all samples prior to the latest date will be considered")
+    parser.add_argument("--latest-date", help="latest date to consider samples for last ancestor or pairwise comparisons (e.g., 2019-01-01); defaults to the current date")
     parser.add_argument("--output", help="JSON file with calculated distances stored by node name and attribute name", required=True)
-    parser.add_argument("--attribute-names", nargs="+", help="names to store distances associated with the corresponding masks", required=True)
-    parser.add_argument("--masks", help="tab-delimited mask definitions with mask name in first column and binary mask in second column")
-    parser.add_argument("--mask-names", nargs="*", help="name of each mask to use from the given masks file. Distances are calculated for each given mask and stored in each of the corresponding attribute names. If no mask is provided, all sites will be used.")
```
Comment (tsibley) [inline augur/distance.py]: This will be a pretty big breaking change. It looks like among our builds the command is only used by seasonal flu? Do we have any reason to think other, external folks are using `augur distance` and would benefit from backwards compat? If we decided it was worth it, It seems like it would be possible to automatically translate the old masks into the new maps and warn about deprecation.

File: augur/distance.py [hunk 4]
```diff
@@ -73,54 +366,99 @@ def run(args):
...
+                file=sys.stderr
+            )
+            sys.exit(1)
+
+        date_annotations = read_node_data(args.date_annotations)
+
+        # Annotate tree with date annotations.
+        for node in tree.find_clades():
+            node.attr = date_annotations["nodes"][node.name]
+            node.attr["num_date"] = node.attr["numdate"]
```
Comment (tsibley) [inline augur/distance.py]: :sob:

File: augur/distance.py [hunk 5]
```diff
@@ -73,54 +366,99 @@ def run(args):
...
+            # previous season.
+            distances_by_node = get_distances_to_all_pairs(
+                tree,
+                sequences_by_node_and_gene,
+                distance_map,
+                earliest_date,
+                latest_date
+            )
+        else:
+            pass
```
Comment (tsibley) [inline augur/distance.py]: I'd argue this should raise an error instead of silently not doing anything.

File: augur/distance.py [hunk 6]
```diff
@@ -73,54 +366,99 @@ def run(args):
...
+                final_distances_by_node[node_name] = {}
+
+            final_distances_by_node[node_name][attribute] = values
+
+    # Prepare params for export.
+    params = {
+        "attribute": args.attribute_name,
+        "compare_to": args.compare_to,
+        "map_name": distance_map_names,
+        "latest_date": args.latest_date
```
Comment (tsibley) [inline augur/distance.py]: Add `earliest_date` also?

File: augur/distance.py [hunk 7]
```diff
@@ -51,18 +343,19 @@ def register_arguments(parser):
     parser.add_argument("--tree", help="Newick tree", required=True)
     parser.add_argument("--alignment", nargs="+", help="sequence(s) to be used, supplied as FASTA files", required=True)
     parser.add_argument('--gene-names', nargs="+", type=str, help="names of the sequences in the alignment, same order assumed", required=True)
+    parser.add_argument("--compare-to", nargs="+", choices=["root", "ancestor", "pairwise"], help="type of comparison between samples in the given tree", required=True)
+    parser.add_argument("--attribute-name", nargs="+", help="name to store distances associated with the given distance map", required=True)
+    parser.add_argument("--map", nargs="+", help="JSON providing the distance map between sites and, optionally, amino acids at those sites", required=True)
```
Comment (tsibley) [inline augur/distance.py]: Can this help describe the map format, or point to a doc which does?

File: augur/distance.py [hunk 8]
```diff
@@ -22,6 +24,296 @@ def read_masks(mask_file):
...
+    0.0
+    """
+    distance_type = type(distance_map["default"])
+    distance = distance_type(0)
+
+    for gene in node_a_sequences:
+        gene_length = len(node_a_sequences[gene])
+
+        for site in range(gene_length):
+            if node_a_sequences[gene][site] != node_b_sequences[gene][site]:
```
Comment (rneher) [inline augur/distance.py]: why not 
```
for site, (a,b) in enumerate(zip(node_a_sequences[gene], node_b_sequences[gene])):
```

**Comment by huddlej** [comment]
I've added more extensive documentation, addressed readability issues, and doublechecked edge cases for date inequalities. All tests pass and the new interface produces appropriate results in auspice when JSONs are built by the `seasonal-flu` repo's `distance-maps` branch.

I can't merge this PR without also merging the corresponding the [seasonal flu PR](https://github.com/nextstrain/seasonal-flu/pull/26). Would someone else (@barneypotter24, @rneher, or @trvrb) mind testing that seasonal flu branch with this augur branch to make sure everything is working?

**Comment by trvrb** [comment]
Thanks John. This is looking really good. @barneypotter24: could you locally test this augur PR against the seasonal flu PR? Just a single end point should be sufficient.

**Comment by barneypotter24** [comment]
I just the live Snakefile using the corresponding `seasonal_flu` branch and it ran/produced output as expected.
